Multiple Displays

robweber opened the pull request:
This is my second draft at this. The first one was way too cumbersome and didn't fit with the focus of this project. This one takes more of a library approach and involves minimal changes to the codebase here. 

## Goal 

Provide a way to allow multiple display types to be used with `slowmovie.py` with the most minimal impact to the developing codebase here. The solution should be user friendly and not require end users to muck about changing python files. 

## Solution 

I've spun off a [separate repo](https://github.com/robweber/vsmp-epd) specifically for abstracting EPD drivers for "vsmp" (very slow movie player) projects. Right now it will handle a mock display (for testing without hardware) and any Waveshare devices in the [Waveshare e-Ink repo](https://github.com/waveshare/e-Paper). Within the next few days I hope to add Inky displays as well, just waiting for my Inky hardware to arrive. 

`vsmp-epd` is published to PyPi and is available as a dependency. This makes it easy to integrate with this project. You can see from the changes to `slowmovie.py` that it's almost a 1:1 replacement in terms of interacting with the library. The main 7.5in display from the instructions is the default but using the `-e` argument you can specify any compatible display. 

## Pros/Cons

The big pro here is this keeps all the E-ink abstraction code separate from this project. The goal here is to provide the media player, not drag along the interaction code for the displays. It also allows for specifying different display types pretty easily for the end user. 

The major con is mainly Waveshare itself. Because it's not available as a `pip` install it will still be a manual build per the install instructions. I can include the Inky libraries as dependencies with `vsmp-epd` but Waveshare will need to have it's own build instructions (unless you're not using that type of display). 

__Fun Note__: I've been running this abstraction code for 2 days now with no issues on my home display.

robweber commented:
Just wanted to __ping__ this PR and see if there are any thoughts on it. Is this a direction anyone is interested in this or are we pretty much just supporting the existing device (hardcoded) and directing people to roll their own via a fork if they need to do something different? 

I've made some pretty good progress on the abstraction library and have basic Inky support hammered out. My plan is to reach out to a few of the known forks where people have done other devices and ask them for a little help in testing to make sure something obvious isn't being missed. Should make this a pretty robust solution.

missionfloyd commented:
Looks like a good way to expand display support without messing with much of the code here.

I do have one (very minor) thought. Why name the module *vsmp*-epd? Surely this is suitable for any e-ink project.

qubist commented on README.md:
````diff
@@ -90,6 +92,17 @@ If an arg is specified in more than one place, then commandline values override
+
+```
+
+You can view a list of compatible e-ink devices by [clicking here](https://github.com/robweber/vsmp-epd/blob/main/README.md#displays-implemented).
````
```suggestion
You can view a list of compatible e-ink devices on the [VSPM-EPD repo](https://github.com/robweber/vsmp-epd/blob/main/README.md#displays-implemented).
```

More context in link text (https://accessibility.psu.edu/linktext/)

qubist commented on slowmovie.py:
```diff
@@ -210,6 +211,17 @@ def estimate_runtime(delay, increment, frames, output="guess"):
+
+if(args.epd not in validEpds):
+    # can't find the driver
+    logging.error(f"{args.epd} is not a valid EPD name, valid names are:")
```
```suggestion
    logging.error(f"'{args.epd}' is not a valid EPD name, valid names are:")
```

I think we've been doing that type of quotes with many of the debugging strings

qubist commented on slowmovie.py:
```diff
@@ -210,6 +211,17 @@ def estimate_runtime(delay, increment, frames, output="guess"):
+    logging.error("\n".join(map(str, validEpds)))
+
+    # can't get past this
+    exit(1)
```
Should this be `sys.exit()` now? I think @missionfloyd standardized the main code to use that a bit ago.

missionfloyd replied:
Yes, `exit()` is for the command line interpreter.

qubist reviewed (approved):
This looks awesome. I think it's definitely a good direction to pursue! I made some minor notes, but I love how it's mainly a drop-in refactor!

One question about the waveshare epd-drivers issue you mentioned: With this PR, the install script would still handle that, right? And if you were using an Inky, for example, you'd just have the install script download the waveshare drivers automatically and then ignore them/delete them if you cared?

qubist commented:
Seems like this would close #62

robweber commented:
Modified per the suggestions above. 

I have been thinking about the library name a bit. @missionfloyd is right - this is generally suited for any e-ink display project where you want to support multiple displays. If a _rebrand_ is a good thought it should be done prior to merging anything into this project since it will affect naming. 

Any thoughts? Something like `universal-epd` or `abstract-epd`? Naming things isn't my strong suit but I think it should be descriptive and not something random like a character name.

qubist commented:
Maybe something like `multi-epd`. The multi-epd library. I think that gets across the essence of the library: supporting multiple e-paper displays in the same library.

missionfloyd commented:
`omni-epd`?

robweber commented:
Both of those are pretty good `omni` sounds pretty cool so I'll run with that. I'll make the necessary changes to the library and then update this PR. So far there is support for: 

* The displays listed in [the table](https://github.com/robweber/vsmp-epd#displays-implemented)
* All Python libraries for the supported EPDs required as part of `setup.cfg` - still need to setup any OS stuff like enabling SPI manually or via scripts from implementing projects (like we have in our `install.sh` script)
* A CLI tool `vsmp-epd-test` - it can write a random image or just a series of squares to your EPD to test driver functionality

I'd like to get more confirmed hardware tests but a problem occurred to me. I can find lots of forks where people have modified code to work with other displays (Inky and Waveshare). The issue is you can't send messages on Github so there isn't any way to reach out and ask them for help testing anything. If anyone has ideas on that front that would be great.

qubist commented:
> The issue is you can't send messages on Github so there isn't any way to reach out and ask them for help testing anything.

Can you add an issue on their fork explaining the situation and asking for testing?

missionfloyd commented:
I could be missing something, but doesn't look like you've accounted for the fact that waveshare 3-color displays [take two images](https://github.com/waveshare/e-Paper/blob/af9694c602fb24d776cd0f0b3e073a477294e3d8/RaspberryPi_JetsonNano/python/lib/waveshare_epd/epd1in54b_V2.py#L134-L150). and they haven't set default values (their drivers could really use some work.)

Seems like [this](https://github.com/robweber/omni-epd/blob/2051ca509dbd905c166ce3b16fc01cc31945672f/src/omni_epd/displays/waveshare_display.py#L66-L67) will fail on them.

missionfloyd commented:
I could be missing something, but it doesn't look like you've accounted for the fact that waveshare 3-color displays [take two images](https://github.com/waveshare/e-Paper/blob/af9694c602fb24d776cd0f0b3e073a477294e3d8/RaspberryPi_JetsonNano/python/lib/waveshare_epd/epd1in54b_V2.py#L134-L150) (their drivers could really use some work.)

robweber commented:
@missionfloyd - Well that's not good. My oversight there. I looked at a good handful of those classes and they all appeared to be the same format. It troubled me a bit they didn't inherit methods from a parent class but thought it was just laziness on their part. Guess there was a reason for that. 

Two options really that I can see:  

1) just remove those devices from the approved list for now
2) create some special handling for those in the abstract class

I think it could be handled fairly well by taking the `Image` passed in and performing some transforms on it. It could be stripped to 3 colors (black/white/red) and then separated again so color 1 went to the black image and color 2 to the red image. The inky library does something similar for their Impression displays to convert an image to 7 colors only.  Without a device to test it on though I'd be a little apprehensive about just writing that code and calling it ready. 

I created [an issue](https://github.com/robweber/omni-epd/issues/7) for this if you want to continue this discussion there on possible solutions.

robweber commented:
After a lot of work in the omni-epd repo @missionfloyd and I have gotten this abstraction library working pretty well. It should handle pretty much any of the waveshare devices as well as the Inky devices; a [full list is here](https://github.com/robweber/omni-epd#displays-implemented). There is also support for different modes on displays that offer more than B/W as a color option. This allows the owners for those types of displays to utilize all the options the display might have without changing any of the code here. 

I did a clean Raspberry Pi OS install and setup all the library requirements from scratch using the instructions in this branch; which are just the normal instructions + adding the omni-epd library. Everything installed great and was working. Before merging this in I'd encourage more testing if possible. Cloning this branch and then running: `sudo pip3 install git+https://github.com/robweber/omni-epd.git#egg=omni-epd` or `sudo pip3 install -r requirements.txt` (with the updated requirements file here) will install the library on your Rpi. For the default Waveshare 7.5in display just running Slowmovie like normal will work. For other displays just pass in the name via the `-e` option. Names are given in the support display list above.

qubist commented:
Would it still help for me to test this even thought I only have the one display?

missionfloyd commented on slowmovie.py:
```diff
@@ -212,6 +211,7 @@ def find_subtitles(file):
 parser.add_argument("-s", "--start", type=int, help="start playing at a specific frame")
 parser.add_argument("-c", "--contrast", default=1.0, type=float, help="adjust image contrast (default: %(default)s)")
 parser.add_argument("-l", "--loop", action="store_true", help="loop a single video; otherwise play through the files in the videos directory")
+parser.add_argument("-e", "--epd", default="waveshare_epd.epd7in5_V2", help="the name of the display device driver to use (default: %(default)s)")
```
```suggestion
parser.add_argument("-e", "--epd", help="the name of the display device driver to use (default: %(default)s)")
```
This is keeping EPD type in `omni-epd.ini` from working. It might make more sense to have the default in the config file, either `slowmovie.conf` or `omni-epd.ini` (maybe those should have the same extension).

But then, if config files are missing, it'll print the list of displays instead.

missionfloyd replied:
It does work the way it is. Might not even be worth messing with.

robweber replied:
So if we leave the `-e` option, but set the default to nothing, we'd have the following order of precedence. 

1. `slowmovie.py` checks for `-e`, if there that's the priority
2. `slowmovie.conf` - essentially the same as checking for `-e` just in the config file
3. `omni-epd.ini` - if nothing else attempting loading from INI file

So if the user deletes `omni-epd.ini`, or otherwise messes with it to screw up the EPD selection __and__ sets no default via the `slowmovie.conf` file or `-e` at the command line they should just get a "no driver found" error. Essentially they'll get a list of options. We could modify the error message to explain what they need to do? We could also re-create some files but if we set things to default and the display is not a 7.5in waveshare there is essentially the same problem of it not working right? I think if we set some defaults in the configs and they get messed with that's not really our fault. We aren't hurting existing installs that may update, to me that's the main thing.

missionfloyd replied:
Sounds good.

missionfloyd commented on slowmovie.py:
```diff
@@ -212,6 +211,7 @@ def find_subtitles(file):
 parser.add_argument("-s", "--start", type=int, help="start playing at a specific frame")
 parser.add_argument("-c", "--contrast", default=1.0, type=float, help="adjust image contrast (default: %(default)s)")
```
Should we keep the contrast argument? Seems redundant now that Omni-EPD handles image adjustments.

robweber replied:
Would be one less arg to carry here. Might make sense to re-work the README a bit to explain values that can be adjusted or at least provide a link to the explanations. I'll work on something for that.

missionfloyd reviewed (commented):
So far, so good on inky impression. I do have some thoughts.

qubist commented:
Ok: cloned the branch, went into `Install` dir, ran `sudo pip3 install -r requirements.txt`, ran the install script for good measure, then ran the test. Works fine on my display. I made sure that the `-e` option existed to validate I had the right version running. One thing I did run in to:

<img width="675" alt="Screen Shot 2021-04-30 at 12 12 29 PM" src="https://user-images.githubusercontent.com/3607279/116723285-57a80e80-a9ad-11eb-8e3c-45bfd17c1e90.png">

I cloned the branch into a differently-named directory (`SlowMovie-Multi`) because I wanted to leave my old one alone. The install script told me to run the wrong file. I'm also seeing now the whiptail menu told me about that:

                     │  Repo set to 'https://github.com/TomWhitwell/SlowMovie/main'     │
                     │  Setting up in local directory '/home/pi/SlowMovie'              │

Maybe I just shouldn't have done this and nothing needs to be changed?

robweber commented:
The install script assumes `/home/pi/SlowMovie` is the directory you're updating. You probably just ended up re-running a pull on your existing install. Usually I'll rename my working directory to `SlowmovieKeep` or something and then clone the branch I'm working on so the install script runs properly. When I'm done I delete it and rename back. Running the `pip` command seems have brought in the required libs though so seems like it's all working OK right?

qubist commented:
@robweber Yeah, I think my run of the install script just did nothing related to this PR. The `pip` command brought everything in fine I believe, although—if it failed to, would I see an error without trying to run things with a non-default screen?

robweber commented:
You'd definitely have problems when trying to load `omni-epd` if things weren't working right. That import would fail right out of the gate before the rest of `slowmovie.py` had a chance to do anything. One thing you could try is just giving it a bogus value - like `-e test`. That should fail out and print a list of supported displays since there isn't a match. 

The test utility `omni-epd-test` is also available as a program from the CLI, running that would confirm the libraries are working. Just run that from anywhere.

missionfloyd commented:
Now that `epd` is in the config file, this is where the display menu I suggested for `install.sh` would come in.

Here's a working implementation using the names from the omni-epd readme.
```shell
DISPLAY=$(whiptail --notags --title "Select your display" --menu "" 0 0 0 \
inky.impression            "Inky Impression 7 Color" \
inky.phat_red              "Inky pHAT Red/Black/White - 212x104" \
inky.phat_yellow           "Inky pHAT Yellow/Black/White - 212x104" \
inky.phat_black            "Inky pHAT Black/White - 212x104" \
inky.phat1608_red          "Inky pHAT Red/Black/White - 250x122" \
inky.phat1608_yellow       "Inky pHAT Yellow/Black/White - 250x122" \
inky.phat1608_black        "Inky pHAT Black/White - 250x122" \
inky.what_red              "Inky wHAT Red/Black/White" \
inky.what_yellow           "Inky wHAT Yellow/Black/White" \
inky.what_black            "Inky wHAT Black/White" \
omni_epd.mock              "Mock Display (emulates EPD with no hardware)" \
waveshare_epd.epdlin02     "Waveshare 1.02inch E-Ink display module" \
waveshare_epd.epdlin54     "Waveshare 1.54inch E-Ink display module" \
waveshare_epd.epdlin54_V2  "Waveshare 1.54inch E-Ink display module V2" \
waveshare_epd.epdlin54b    "Waveshare 1.54inch e-Paper Module B" \
waveshare_epd.epdlin54b_V2 "Waveshare 1.54inch e-Paper Module B V2" \
waveshare_epd.epdlin54c    "Waveshare 1.54inch e-Paper Module C" \
waveshare_epd.epd2in13     "Waveshare 2.13inch e-Paper HAT" \
waveshare_epd.epd2in13_V2  "Waveshare 2.13inch e-Paper HAT V2" \
waveshare_epd.epd2in13b    "Waveshare 2.13inch e-Paper HAT B" \
waveshare_epd.epd2in13b_V3 "Waveshare 2.13inch e-Paper HAT B V3" \
waveshare_epd.epd2in13c    "Waveshare 2.13inch e-Paper HAT C" \
waveshare_epd.epd2in13d    "Waveshare 2.13inch e-Paper HAT D" \
waveshare_epd.epd2in66     "Waveshare 2.66inch e-Paper Module" \
waveshare_epd.epd2in66b    "Waveshare 2.66inch e-Paper Module B" \
waveshare_epd.epd2in7      "Waveshare 2.7inch e-Paper HAT" \
waveshare_epd.epd2in7b     "Waveshare 2.7inch e-Paper HAT B" \
waveshare_epd.epd2in7b_V2  "Waveshare 2.7inch e-Paper HAT B V2" \
waveshare_epd.epd2in9      "Waveshare 2.9inch e-Paper Module" \
waveshare_epd.epd2in9_V2   "Waveshare 2.9inch e-Paper Module V2" \
waveshare_epd.epd2in9b     "Waveshare 2.9inch e-Paper Module B" \
waveshare_epd.epd2in9b_V3  "Waveshare 2.9inch e-Paper Module B V3" \
waveshare_epd.epd2in9c     "Waveshare 2.9inch e-Paper Module C" \
waveshare_epd.epd2in9d     "Waveshare 2.9inch e-Paper HAT D" \
waveshare_epd.epd3in7      "Waveshare 3.7inch e-Paper HAT" \
waveshare_epd.epd4in01f    "Waveshare 4.01inch 7 color e-Paper HAT" \
waveshare_epd.epd4in2      "Waveshare 4.2inch e-Paper Module" \
waveshare_epd.epd4in2b     "Waveshare 4.2inch e-Paper Module B" \
waveshare_epd.epd4in2b_V2  "Waveshare 4.2inch e-Paper Module B V2" \
waveshare_epd.epd4in2c     "Waveshare 4.2inch e-Paper Module C" \
waveshare_epd.epd5in65f    "Waveshare 5.65inch e-Paper Module F" \
waveshare_epd.epd5in83     "Waveshare 5.83inch e-Paper HAT" \
waveshare_epd.epd5in83_V2  "Waveshare 5.83inch e-Paper HAT V2" \
waveshare_epd.epd5in83b    "Waveshare 5.83inch e-Paper HAT B" \
waveshare_epd.epd5in83b_V2 "Waveshare 5.83inch e-Paper HAT B V2" \
waveshare_epd.epd5in83c    "Waveshare 5.83inch e-Paper HAT C" \
waveshare_epd.epd7in5      "Waveshare 7.5inch e-Paper HAT" \
waveshare_epd.epd7in5_V2   "Waveshare 7.5inch e-Paper HAT V2" \
waveshare_epd.epd7in5_HD   "Waveshare 7.5inch HD e-Paper HAT" \
waveshare_epd.epd7in5b_HD  "Waveshare 7.5inch HD e-Paper HAT B" \
waveshare_epd.epd7in5b     "Waveshare 7.5inch e-Paper HAT B" \
waveshare_epd.epd7in5b_V2  "Waveshare 7.5inch e-Paper HAT B V2" \
waveshare_epd.epd7in5c     "Waveshare 7.5inch e-Paper HAT C" \
3>&2 2>&1 1>&3)

if [ $? == 0 ]; then
  grep -q "^epd[[:blank:]]*=[[:blank:]]*" /home/pi/SlowMovie/slowmovie.conf 2>/dev/null && sed -i "/^epd\s*=\s*/s/.*/epd = $DISPLAY/" /home/pi/SlowMovie/slowmovie.conf || echo "epd = $DISPLAY" >> /home/pi/SlowMovie/slowmovie.conf
fi
```
This will update the `epd` setting in `slowmovie.conf`, or add it if it missing.

This even could be taken a step further by running `omni-epd-test` with the selected display, and asking the user if it worked. But that might be overkill.

robweber commented:
I moved the default waveshare device to the conf file and removed the default for `-e` in the main program. This should allow the user to use basically any combination of the conf file, CLI, or omni-epd.ini file to set their default. Also updated the README to add the other INI file values that could be used. I left the contrast option for now, we can always remove in another PR just to keep this from modifying too many pieces.

robweber commented:
Looking through this I _think_ we're ok. All major issues should be addressed.

qubist commented:
Yup, looks good!
